src/istgt_lu_disk.c

```diff
@@ -5552,68 +5564,214 @@ istgt_lu_disk_scsi_reserve(ISTGT_LU_DISK *spec, CONN_Ptr conn, ISTGT_LU_CMD_Ptr
+		struct zvol_io_rw_hdr *rio_rw_hdr =			\
+		    (struct zvol_io_rw_hdr *)(ldata +			\
+		    sizeof(zvol_io_hdr_t));				\
+		rcmd = get_from_mempool(&rcmd_mempool);			\
+		memset(rcmd, 0, sizeof(*rcmd));				\
+		rcmd->opcode = rcomm_cmd->opcode;			\
+		rcmd->offset = rcomm_cmd->offset;			\
+		rcmd->data_len = rcomm_cmd->data_len;			\
+		rcmd->io_seq = rcomm_cmd->io_seq;			\
+		rcmd->idx = rcomm_cmd->copies_sent - 1;			\
+		rcmd->healthy_count = spec->healthy_rcount;		\
+		rcmd->rcommq_ptr = rcomm_cmd;				\
+		rcmd->status = 0;					\
+		rcmd->iovcnt = rcomm_cmd->iovcnt;			\
+		for (i=1; i < rcomm_cmd->iovcnt + 1; i++) {		\
+			rcmd->iov[i].iov_base = 			\
+			     rcomm_cmd->iov[i].iov_base;		\
+			rcmd->iov[i].iov_len = 				\
+			    rcomm_cmd->iov[i].iov_len;			\
+		}							\
+		rio->opcode = rcmd->opcode;				\
+		rio->version = REPLICA_VERSION;				\
+		rio->io_seq = rcmd->io_seq;				\
+		rio->offset = rcmd->offset;				\
+		if (rcmd->opcode == ZVOL_OPCODE_WRITE) {		\
+			rio->len = rcmd->data_len +			\
+			    sizeof(struct zvol_io_rw_hdr);		\
+			rio->checkpointed_io_seq = 0;			\
+		} else							\
+			rio->len = rcmd->data_len;			\
+		rcmd->iov_data = ldata;					\
+		rio_rw_hdr->io_num = rcmd->io_seq;			\
+		rio_rw_hdr->len = rcmd->data_len;			\
+		rcmd->iov[0].iov_base = rio;				\
+		if (rcomm_cmd->opcode == ZVOL_OPCODE_WRITE)		\
+			rcmd->iov[0].iov_len = sizeof(zvol_io_hdr_t) +	\
+			    sizeof(struct zvol_io_rw_hdr);		\
+		else							\
+			rcmd->iov[0].iov_len = sizeof(zvol_io_hdr_t);	\
+		rcmd->iovcnt++;						\
+	} while (0);							\
 
 void
 clear_rcomm_cmd(rcommon_cmd_t *rcomm_cmd)
 {
 	int i;
-	pthread_mutex_destroy(&(rcomm_cmd->rcommand_mtx));
 	for (i=1; i<rcomm_cmd->iovcnt + 1; i++)
 		xfree(rcomm_cmd->iov[i].iov_base);
 	put_to_mempool(&rcommon_cmd_mempool, rcomm_cmd);
 }
 
+//extern void get_all_read_resp_data_chunk(void *data, struct io_data_chunk_list_t *io_data_chunk);
+//extern uint8_t * process_chunk_read_resp(struct io_data_chunk_list_t  *io_chunk_list, uint64_t len);
+
+static uint8_t *
+handle_read_consistency(rcommon_cmd_t *rcomm_cmd)
+{
+	int i;
+	uint8_t *dataptr = NULL;
+	struct io_data_chunk_list_t io_data_chunk_list;
+
+	TAILQ_INIT(&(io_data_chunk_list));
+
+	for (i = 0; i < rcomm_cmd->copies_sent; i++) {
+		if (rcomm_cmd->resp_list[i].status == 1) {
+			get_all_read_resp_data_chunk(&rcomm_cmd->resp_list[i], &io_data_chunk_list);
+		}
+	}
+
+	dataptr = process_chunk_read_resp(&io_data_chunk_list, rcomm_cmd->data_len);
+
+	return dataptr;
+}
+
+static int
+check_for_command_completion(spec_t *spec, rcommon_cmd_t *rcomm_cmd, ISTGT_LU_CMD_Ptr cmd)
+{
+	int i, rc = 0;
+	uint8_t *data = NULL;
+	int success = 0, failure = 0;
+
+	for (i = 0; i < rcomm_cmd->copies_sent; i++) {
+		if (rcomm_cmd->resp_list[i].status == 1) {
+			success++;
+		} else if (rcomm_cmd->resp_list[i].status == -1) {
+			failure++;
+		}
+	}
+
+	if (rcomm_cmd->opcode == ZVOL_OPCODE_READ) {
+		if ((rcomm_cmd->copies_sent != (success + failure))) {
```

Comment:
we don't need to wait till all responses comes. If 'success' is reached to required number (i.e, either one success response from healthy replica or (n-m+1) success responses from degraded.. here, n is replication factor, m is consistency factor), response can be sent back to client.